usim/_primitives/context.py

```diff
@@ -155,46 +219,80 @@ def _close_volatile(self):
         return self
 
     async def __aexit__(self, exc_type, exc_val, exc_tb):
-        # receiving GeneratorExit means our coroutine is .closed'd
-        # This is forceful, and we are not allowed to await anything
-        # since it can happen on any await, we check multiple times
-        if exc_type is GeneratorExit:
-            return self._handle_close(exc_val)
-        assert (
-            self._activity is __LOOP_STATE__.LOOP.activity
-        ), (
-            "Instances of %s cannot be shared between activities" %
-            self.__class__.__name__
-        )
-        await self._done.set()
         if exc_type is None:
             try:
-                await self._await_children()
+                # inform everyone that we are shutting down
+                # we may receive any shutdown signal here
+                await self._body_done.set()
             except BaseException as err:
                 exc_type, exc_val = type(err), err
             else:
-                self._close_volatile()
-                return self._handle_exception(exc_val)
+                return await self._aexit_graceful()
+        self._body_done._value = True
+        self._body_done.__trigger__()
         # there was an exception, we have to abandon the scope
-        if exc_type is GeneratorExit:
-            return self._handle_close(exc_val)
+        return self._aexit_forceful(exc_type, exc_val)
+
+    def _aexit_forceful(self, exc_type, exc_val):
+        """
+        Exit with exception
+
+        This immediately closes all children without waiting for anything.
+        No further interrupts can occur during shutdown (this is a sync function).
+
+        If a fatal exception occurred in this scope or a child, the fatal
+        exception replaces all other exceptions in flight.
+        If the current exception is suppressed by the scope, any exceptions from
+        children are reraised as :py:exc:`~.Concurrent` errors.
+        """
+        # we already handle an exception, suppress
+        self._disable_interrupts()
         # reap all children now
-        self._cancel_children()
-        await self._await_children()
-        if exc_type is GeneratorExit:
-            return self._handle_close(exc_val)
+        self._close_children()
         self._close_volatile()
-        return self._handle_exception(exc_val)
+        # allow replacing our exception with more important ones
+        if not issubclass(exc_type, self.PROMOTE_CONCURRENT):
+            self._reraise_privileged()
+        if self._suppress_exception(exc_val):
+            self._reraise_concurrent()
```

Comment:
Can it happen that privileged are reraised again as this was also included in `reraise_privileged`?

Reply:
Given the current ``PROMOTE_CONCURRENT``, there is no situation where an exception is privileged but would be suppressed. Not sure whether a subclass would want to change that.